src/interface/sparse_pack_base.cpp

```diff
@@ -0,0 +1,265 @@
+          if (pv->IsAllocated()) {
+            for (int t = 0; t < pv->GetDim(6); ++t) {
+              for (int u = 0; u < pv->GetDim(5); ++u) {
+                for (int v = 0; v < pv->GetDim(4); ++v) {
+                  if (pack.coarse_) {
+                    pack_h(0, b, idx) = pv->coarse_s.Get(t, u, v);
+                  } else {
+                    pack_h(0, b, idx) = pv->data.Get(t, u, v);
+                  }
+                  PARTHENON_REQUIRE(
+                      pack_h(0, b, idx).size() > 0,
+                      "Seems like this variable might not actually be allocated.");
+                  if (desc.with_fluxes && pv->IsSet(Metadata::WithFluxes)) {
+                    pack_h(1, b, idx) = pv->flux[1].Get(t, u, v);
+                    PARTHENON_REQUIRE(pack_h(1, b, idx).size() ==
+                                          pack_h(0, b, idx).size(),
+                                      "Different size fluxes.");
+                    if (ndim > 1) {
+                      pack_h(2, b, idx) = pv->flux[2].Get(t, u, v);
+                      PARTHENON_REQUIRE(pack_h(2, b, idx).size() ==
+                                            pack_h(0, b, idx).size(),
+                                        "Different size fluxes.");
+                    }
+                    if (ndim > 2) {
+                      pack_h(3, b, idx) = pv->flux[3].Get(t, u, v);
+                      PARTHENON_REQUIRE(pack_h(3, b, idx).size() ==
+                                            pack_h(0, b, idx).size(),
+                                        "Different size fluxes.");
+                    }
+                  }
+                  idx++;
+                }
+              }
+            }
+          }
+        }
+      }
+
+      bounds_h(1, b, i) = idx - 1;
+
+      if (bounds_h(1, b, i) < bounds_h(0, b, i)) {
+        // Did not find any allocated variables meeting our criteria
+        bounds_h(0, b, i) = -1;
+        // Make the upper bound more negative so a for loop won't iterate once
+        bounds_h(1, b, i) = -2;
+      }
+    }
+  });
+
+  Kokkos::deep_copy(pack.pack_, pack_h);
+  Kokkos::deep_copy(pack.bounds_, bounds_h);
+  Kokkos::deep_copy(pack.coords_, coords_h);
+  pack.ndim_ = ndim;
+  pack.dims_[1] = pack.nblocks_;
+  pack.dims_[2] = -1; // Not allowed to ask for the ragged dimension anyway
+  pack.dims_[3] = pack_h(0, 0, 0).extent_int(0);
+  pack.dims_[4] = pack_h(0, 0, 0).extent_int(2);
+  pack.dims_[5] = pack_h(0, 0, 0).extent_int(3);
+
+  return pack;
+}
+
+// Specialize for the only two types this should work for
+template SparsePackBase
+SparsePackBase::Build<MeshBlockData<Real>>(MeshBlockData<Real> *, const PackDescriptor &);
+template SparsePackBase SparsePackBase::Build<MeshData<Real>>(MeshData<Real> *,
+                                                              const PackDescriptor &);
+
+inline SparsePackBase::test_func_t
+SparsePackBase::GetTestFunction(const PackDescriptor &desc) {
```

Comment:
Is there a good reason to create a test function rather than just having a test function? To me, it would make more sense for the `PackDescriptor` to have such a test function built in.

Also, "GetTestFunction" is an ambiguous name, it doesn't suggest that it returns a function that tests whether a `CellVariable` is included by the provided `PackDescriptor`.

Reply:
I have to think about it a little bit more, but I think you are right that this could just be a member function of `PackDescriptor`. `GetTestFunction` predates `PackDescriptor` in the development process, so there may be no good reason for it anymore. It was just there to easily create a reusable functor and reduce code duplication.

Reply:
Yep, your suggestion is definitely cleaner. I have made the change.